components/merino_client/src/client.rs

```diff
@@ -0,0 +1,381 @@
+    fn from(suggestion: SuggestResponseSuggestion) -> Self {
+        match suggestion {
+            SuggestResponseSuggestion::Known(SuggestResponseKnownProviderSuggestion::Adm {
+                details,
+                block_id,
+                full_keyword,
+                advertiser,
+                impression_url,
+                click_url,
+            }) => MerinoSuggestion::Adm {
+                details,
+                block_id,
+                full_keyword,
+                advertiser,
+                impression_url,
+                click_url,
+            },
+            SuggestResponseSuggestion::Known(
+                SuggestResponseKnownProviderSuggestion::TopPicks {
+                    details,
+                    block_id,
+                    is_top_pick,
+                },
+            ) => MerinoSuggestion::TopPicks {
+                details,
+                block_id,
+                is_top_pick,
+            },
+            SuggestResponseSuggestion::Unknown { details, provider } => {
+                MerinoSuggestion::Other { details, provider }
+            }
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct SuggestResponse {
+    suggestions: Vec<SuggestResponseSuggestion>,
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum SuggestResponseSuggestion {
+    Known(SuggestResponseKnownProviderSuggestion),
+    // `#[serde(other)]` doesn't support associated data, so we can't
+    // deserialize the response suggestion directly into a `MerinoSuggestion`.
+    // Instead, we have an "outer", untagged `SuggestResponseSuggestion` that
+    // has our unknown / other variant, and an "inner", internally tagged
+    // `SuggestResponseKnownProviderSuggestion` with our known variants.
+    Unknown {
+        #[serde(flatten)]
+        details: MerinoSuggestionDetails,
+        provider: String,
+    },
+}
+
+#[derive(Deserialize)]
+#[serde(tag = "provider")]
+enum SuggestResponseKnownProviderSuggestion {
+    #[serde(rename = "adm")]
+    Adm {
+        #[serde(flatten)]
+        details: MerinoSuggestionDetails,
+        block_id: i64,
+        full_keyword: String,
+        advertiser: String,
+        impression_url: Option<String>,
+        click_url: Option<String>,
+    },
+    #[serde(rename = "top_picks")]
+    TopPicks {
+        #[serde(flatten)]
+        details: MerinoSuggestionDetails,
+        block_id: i64,
+        is_top_pick: bool,
+    },
+}
+
+struct SessionState {
+    started_at: Instant,
+    session_id: String,
+    sequence_number: i64,
+}
+
+impl Default for SessionState {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SessionState {
+    fn new() -> Self {
+        let mut session_id_bytes = vec![0u8; 16];
+        rand::fill(&mut session_id_bytes).unwrap();
+        Self {
+            started_at: Instant::now(),
+            session_id: hex::encode(&session_id_bytes),
+            sequence_number: 0,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
suggestion (blocking): Let's add some more tests! Ideas for test cases:

* Other kinds of suggestions (Top Picks, AccuWeather).
* Specifying additional settings, like the default providers, client variants.
* Explicitly specifying `providers` when calling `fetch`.
* Session ID and sequence number rotation.
* Error responses (non-200 response code, bad JSON in body).

/cc @tiftran